js/lib/SparseMerkleTree.js

```diff
@@ -0,0 +1,147 @@
+const { sha3, sha256, bufferToHex } = require('ethereumjs-util');
+const BigNumber = require('bignumber.js');
+
+const hash = sha3;
+
+class SparseMerkleTree {
+  constructor(input, depth) {
+    this.depth = depth || 256;
+    const len = Object.keys(input).length;
+    const max = Math.pow(2, this.depth - 1);
+    if (len > max) {
+      throw new Error('There are too many leaves for the tree to build');
+    }
+
+    // convert indices to HEX strings
+    const hexInput = {};
+    for (let index in input) {
+      if (index < 0) {
+        throw new Error('Cannot parse negative indices')
+      }
+      try {
+        hexInput[add(index, 0)] = input[index];
+      } catch (err) {
+        throw new Error('Cannot parse an index: ' + index)
+      }
+
+      // TODO: check if indexes within [0, max)
+    }
+
+    this.defaultNodes = SparseMerkleTree.getDefaultNodes(this.depth);
+
+    if (len > 0) {
+      this.tree = this.createTree(hexInput, this.depth);
+      this.root = Object.values(this.tree.slice(-1)[0]).slice(-1)[0];
+    } else {
+      this.tree = [];
+      this.root = this.defaultNodes.slice(-1)[0];
+    }
+  }
+
+  getProofForIndex(idx) {
+    idx = add(idx, 0);
+    const proof = [];
+    for (let level = 0; level < (this.depth - 1); ++level) {
+      const siblingIndex = isEven(idx) ? add(idx, 1) : add(idx, -1);
+      idx = div2(idx);
+      if (siblingIndex in this.tree[level]) {
+        proof.push(this.tree[level][siblingIndex]);
+      } else {
+        proof.push(this.defaultNodes[level]);
+      }
+    }
+    return proof;
+  }
+
+  getHexProofForIndex(idx) {
+    const proof = this.getProofForIndex(idx);
+
+    return SparseMerkleTree.bufArrToHex(proof);
+  }
+
+  getRoot() {
+    return this.root;
+  }
+
+  getHexRoot() {
+    return bufferToHex(this.getRoot());
+  }
+
+  verifyProof(proof, root, leaf, idx) {
+    idx = add(idx, 0);
+    let computedHash = leaf;
+    proof.forEach(function (proofElement) {
```

Comment:
`proof.forEach(proofElement => {`. also should we consider using lodash? this will probably get rid of all the `add`, `even` functions that you have as well